authors/apps/authentication/views.py

```diff
@@ -151,3 +159,71 @@ def put(self, request, token, **kwargs):
         except jwt.ExpiredSignatureError:
             return Response({"The link expired"},
                             status=status.HTTP_400_BAD_REQUEST)
+
+
+class SocialAuthenticationView(generics.CreateAPIView):
+    """Social authentication."""
+    permission_classes = (AllowAny,)
+    serializer_class = SocialAuthSerializer
+    render_classes = (UserJSONRenderer,)
+
+    def create(self, request, *args, **kwargs):
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        provider = request.data['provider']
+        strategy = load_strategy(request)
+        authenticated_user = request.user if not request.user.is_anonymous else None
+
+        try:
+            backend = load_backend(
+                strategy=strategy,
+                name=provider,
+                redirect_uri=None
+            )
+
+            if isinstance(backend, BaseOAuth1):
+                if "access_token_secret" in request.data:
+                    token = {
+                        'oauth_token': request.data['access_token'],
+                        'oauth_token_secret':
+                        request.data['access_token_secret']
+                    }
+                else:
+                    return Response({'error':
+                                    'Please enter your secret token'},
+                                    status=status.HTTP_400_BAD_REQUEST)
+            elif isinstance(backend, BaseOAuth2):
+                token = serializer.data.get('access_token')
+        except MissingBackend:
+            return Response({
+                'error': 'Please enter a valid social provider'
+                }, status=status.HTTP_400_BAD_REQUEST)
+
+        try:
+            user = backend.do_auth(token, user=authenticated_user)
+        except (AuthAlreadyAssociated, IntegrityError):
+            return Response({
+                "errors": "User is authenticated"},
+                status=status.HTTP_400_BAD_REQUEST)
+        except BaseException:
+            return Response({
+                "errors": "The token has expired"},
+                 status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+        if user:
+            user.is_active = True
+            username = user.username
```

Comment:
What happens if username or email already exists. You have to handle it.

Reply:
I have handled this, already